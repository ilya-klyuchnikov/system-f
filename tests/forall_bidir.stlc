(   (lam ((f (-> int int)))
      (f 5))
    (lam (x) (- x 1))         )

(let ((id (Lam (a) (lam (x) x))
          (all (a) (-> a _))))
  id)

(let ((id2 (Lam (a) (lam ((x a)) x))))
  id2)

(let ((ap (Lam (p) (lam (x) (Lam (q) (lam (f) (f x)))))
          (all (a) (-> a (all (b) (-> (-> a b) b))))))
  ap)

(let ((ap (Lam (p) (lam (x) (Lam (q) (lam ((f (-> p q))) (f x)))))
          (all (a) (-> a (all (b) _)))))
  ap)
