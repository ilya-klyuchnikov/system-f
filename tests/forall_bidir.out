>  : int
-> 4
>  : (all (a) (-> a a))
-> #<function>
>  : (all (a) (-> a a))
-> #<function>
>  : (all (a) (-> a (all (a0) (-> (-> a a0) a0))))
-> #<function>
>  : (all (a) (-> a (all (a0) (-> (-> a a0) a0))))
-> #<function>
> 